### PR TITLE
Allow tracking backend events

### PR DESCRIPTION
**Pull request by brunvez**
### Context:

We love the extension and use it every day, but we find ourselves lacking a similar tool to test server-side events since the [Segment Debugger](https://segment.com/docs/connections/sources/debugger/) has its shortcomings; like limiting the number of events shown or not being able to perform a granular search.

### Proposed solution:

This PR adds support for server-side events by reading the `X-Tracked-Events` response header. The idea is that the server will be responsible for sending a JSON object with all the tracked events in that header in a certain format and the extension will then read it and display it.

The usage of this feature is mostly for development and QA, not for production. It also has the con of not working with events sent after the response has been sent from the server, so async events are still difficult to track. The good side is that it's lightweight and platform agnostic, so any technology capable of sending custom response headers should be able to use it.

This is a first draft, so open to suggestions. Also, I tried to keep it optional and it will go unnoticed by people currently using the extension but totally okay if you think this belongs in another extension.

image
image

### CODE REVIEW

File: background.js [hunk 1]
```diff
@@ -19,23 +19,40 @@ function zeroPad(i) {
...
+		if (tab) {
+			callback(tab);
+		}
+	});
+}
+
 function updateTrackedEventsForTab(tabId,port) {
 	var sendEvents = [];
-	
+
```
Comment (AlexeyMK) [inline background.js:41]: probably a formatter issue here, right? worth running through a linter so we're not committing random whitespace

File: background.js [hunk 2]
```diff
@@ -62,55 +79,57 @@ function isSegmentApiCall(url) {
 	return apiDomainParts.findIndex(d => url.startsWith(`https://${d.trim()}`)) != -1;
 }
 
+function onOwnServerpiCall(url, callback) {
```
Comment (AlexeyMK) [inline background.js:82]: nit: typo - 
```suggestion
function onOwnServerApiCall(url, callback) {
```

File: background.js [hunk 3]
```diff
@@ -62,55 +79,57 @@ function isSegmentApiCall(url) {
...
 				}
 				else if (details.url.endsWith('/v1/batch') || details.url.endsWith('/v2/batch') || details.url.endsWith('/v1/b') || details.url.endsWith('/v2/b')) {
-					event.eventName = 'Batch';
 					event.type = 'batch';
-					
+				}
+
+				if (event.type) {
+					event.eventName = eventTypeToName(event.type) || rawEvent.event
+					console.log(event.type, eventTypeToName(event.type) || rawEvent.event, event)
```
Comment (AlexeyMK) [inline background.js:132]:
```suggestion
```
left from debugging?

File: background.js [hunk 4]
```diff
@@ -123,3 +142,34 @@ chrome.webRequest.onBeforeRequest.addListener(
 	]},
 	['blocking','requestBody']
 );
+
+chrome.webRequest.onHeadersReceived.addListener(
+	(details) => {
+		onOwnServerpiCall(details.url, () => {
```
Comment (AlexeyMK) [inline background.js:148]:
```suggestion
		onOwnServerApiCall(details.url, () => {
```

File: background.js (hunk 3 above)
Comment (brunvez) [reply background.js:132]: 👀

**Comment by MartinMouritzen** [comment]
I don't have much  experience with serverside events, but it looks good to me, so I'll merge :)

**Pull request merged**